src/module.c

```diff
@@ -2735,16 +2759,104 @@ int RM_StringTruncate(RedisModuleKey *key, size_t newlen) {
 /* --------------------------------------------------------------------------
  * ## Key API for List type
  *
+ * Many of the list functions access elements by index. Since a list is in
+ * essence a doubly-linked list, accessing elements by index is generally an
+ * O(N) operation. However, if elements are accessed sequentially or with
+ * indices close together, the functions are optimized to seek the index from
+ * the previous index, rather than seeking from the ends of the list.
+ *
+ * This enables iteration to be done efficiently using a simple for loop:
+ *
+ *     long n = RM_ValueLength(key);
+ *     for (long i = 0; i < n; i++) {
+ *         RedisModuleString *elem = RedisModule_ListGet(key, i);
+ *         // Do stuff...
+ *     }
+ *
+ * Note that after modifying a list using RM_ListPop, RM_ListSet or
+ * RM_ListInsert, the internal iterator is invalidated so the next operation
+ * will require a linear seek.
```

Comment:
There is a hidden assumption here that only `RM_List*` functions can mutate the list, but there are other options where it can happen indirectly (`RM_Call()` for example). I think it may be a good idea to add a warning about this here.

Reply:
Good point! RM_Call can confuse the internal iterator. I will add it tomorrow. Do you know any other ways to mutate a list from a module?

Reply:
Can't think of other legitimate use cases, there are probably a few twisted ones --

* Opening the same key several times concurrently
* Holding an open key beyond the callback scope / module GIL lock in a thread safe context

But I guess these scenarios will not end well even before this change, so we may not need to even mention that.

Reply:
so what did we decide here? enhance the warning? or shall i merge as is?

Reply:
Oh, ready to merge? Then I'll edit the warning quickly. It'll be ready to merge today.

Reply:
I think mixing RM_Call with API in not unlikely since there are commands for searching by occurrence etc. which are probably faster with RM_Call than with the API (no dup for each string). So that warning can be quite good to have.

Reply:
OK, I added a warning about RM_Call. @oranagra Feel free to merge.